Doc revamp and cleanup (#2806)
dfinity/motoko: doc/modules/language-guide/pages/motoko-introduction.adoc

:sdk-short-name: DFINITY Canister SDK
:sdk-long-name: DFINITY Canister Software Development Kit (SDK)

> {proglang} is a modern, general-purpose programming language you can use specifically to author link:../developers-guide/concepts/what-is-ic{outfilesuffix}#ic-overview[Internet Computer]  canister smart contracts.

Reviewer: s/author link/link ?

Author: Not sure what you wanted here? Maybe make it as a suggestion? @osa1

Reviewer: Sorry, this looks good, I just misread it. Closing.

> Although aimed squarely at the IC, its design is general enough to support future compilation to other targets.

Author: ```suggestion
Although aimed squarely at the {IC}, its design is general enough to support future compilation to other targets.
```

== Approachability

{proglang} is a modern language designed to be approachable for programmers who have some basic familiarity with modern object-oriented and/or functional programming idioms in either JavaScript, or another modern programming language, such as Rust, Swift, TypeScript, C#, or Java.

== Asynchronous messaging and type sound execution

{proglang} permits modern programming idioms, including special programming abstractions for distributed applications (dapps).

> Each dapp consists of one or more  _actor_s that communicates _without_ using shared state, but instead by using (asynchronous) message passing.

Author: ```suggestion
Each dapp consists of one or more  _actor_s that communicate solely by asynchronous message passing. The state of an actor is isolated from all other actors, supporting distribution. There is no way to share state between several actors. Each  message to an actor is queued and processed in sequence, preventing unexpected race conditions. 
```

Specifically, {proglang} programs are _type sound_ since {proglang} includes a practical, modern type system that checks each one before it executes.
The {proglang} type system statically checks that each {proglang} program will execute safely, without dynamic type errors, on all possible inputs.
Consequently, entire classes of common programming pitfalls that are common in other languages, and web programming languages in particular, are ruled out. This includes null reference errors, mis-matched argument or result types, missing field errors and many others.

To execute, {proglang} statically compiles to link:about-this-guide{outfilesuffix}#wasm[WebAssembly], a portable binary format that abstracts cleanly over modern computer hardware, and thus permits its execution broadly on the Internet, and the link:../developers-guide/concepts/what-is-ic{outfilesuffix}#ic-overview[{IC}].

[[pitch-actors]]
== Each canister smart constract as an _actor_

{proglang} provides an *actor-based* programming model to developers to express _services_, including those of canister smart contracts on the link:../developers-guide/concepts/what-is-ic{outfilesuffix}#ic-overview[{IC}].

> An actor is similar to an object, but is special in that it state is completely isolated, and all its interactions with the world are by _asynchronous_ messaging.

Reviewer: s/it/its

Author: ```suggestion
An actor is similar to an object, but is special in that its state is completely isolated, and all its interactions with the world are by _asynchronous_ messaging.
```

All communication with and between actors involves passing messages asynchronously over the network using the Internet Computer's messaging protocol.
An actor’s messages are processed in sequence, so state modifications never admit race conditions (unless explicitly allowed by punctuating `await`s).

The Internet Computer ensures that each message that is sent receives
a response. The response is either success with some value, or
an error. An error can be the explicit rejection of the message by the
receiving canister, a trap due to an illegal instruction such as division by zero, or a system error due to distribution or resource constraints.
For example, a system error might be
the transient or permanent unavailability of the receiver (either because

> the receiving actor is oversubscribed or has been deleted).

Reviewer: Not a user-visible thing but this paragraph seems to be split into multiple lines randomly (some lines long, others short). Maybe merge this as a single line? If you use vim, you can also select the text and do `gq` to automatically format it based on your `textwidth`.

Author: Done

We can summarize the program's behavior with three steps:

. The program makes two requests (lines 1 and 2) to two

> distinct services, each implemented as a Motoko actor or canister smart contract implemented in some other language.

Author: ```suggestion
distinct services, each implemented as a {proglang} actor or canister smart contract implemented in some other language.
```

In an _asynchronous_ computing setting, a program and its running environment are permitted to perform _internal computations_ that occur _concurrently_ with one another.

Specifically, asynchronous programs are ones where the program's requests of its environment do not (necessarily) require the program to wait for the environment.

> In the meantime, the program is permitted to make internal progress within this environment while it waits, perhaps _actively_. In the example, above, the program issues the second request before waiting for the first request to complete.

Reviewer: I don't understand what actively waiting means, but probably not important for this PR..

Author: YEah, not my text. maybe, "while the environment proceeds with completing the request."

Author: ```suggestion
In the meantime, the program is permitted to make internal progress within this environment while the environment proceeds to complete the request. In the example, above, the program issues the second request before waiting for the first request to complete.
```

> Symmetrically, the environment's requests of the program do not (necessarily) require the environment to wait for the program's answer to make external progress around it while it waits for this answer.

Author: ```suggestion
Symmetrically, the environment's requests of the program do not (necessarily) require the environment to wait for the program's answer: the environment can make external progress while the answer is produced.
```
